build: add support to the rollup build for building typescript packages

michaelfaith opened the pull request:
Contributing to #32240 this change adds support to the rollup build script for building packages with typescript source.  In order for a bundle to be picked up by the new flow, the bundle definition should have a `tsconfig` property with a path to the project's tsconfig.

There's also two new bundle types that such packages can request to be built: `ESM_DTS` and `CJS_DTS`, which will create the corresponding `.d.ts` bundle.

poteto commented on scripts/rollup/build.js:
```diff
@@ -368,27 +378,36 @@ function getPlugins(
-          };
-        },
-      },
+      bundle.tsconfig
```
```suggestion
      bundle.tsconfig != null
```

poteto replied:
Let's keep any comparisons as specific as possible

michaelfaith replied:
Updated

poteto reviewed (changes requested):
I took a look at the [built files](https://github.com/facebook/react/actions/runs/13349841708/artifacts/2597973523) and I noticed that the d.ts files were missing, could you check and see?

michaelfaith commented:
> I took a look at the [built files](https://github.com/facebook/react/actions/runs/13349841708/artifacts/2597973523) and I noticed that the d.ts files were missing, could you check and see?

This is because we haven't added the `tsconfig` and updated `target` prop to the `eslint-plugin-react-hooks` bundle definition yet, which we'll have to do in the main branch where we've converted the plugin to TS.  This just prepares the build for that.  In the main PR, we'll just need to do the following in `bundles.js`.

```diff
{
+   name: 'eslint-plugin-react-hooks',
-   bundleTypes: [NODE_DEV, NODE_PROD],
+   bundleTypes: [NODE_DEV, NODE_PROD, CJS_DTS],
    moduleType: ISOMORPHIC,
-   entry: 'eslint-plugin-react-hooks',
+   entry: 'eslint-plugin-react-hooks/src/index.ts',
    global: 'ESLintPluginReactHooks',
    minifyWithProdErrorCodes: false,
    wrapWithModuleBoundaries: false,
    externals: [],
+   tsconfig: './packages/eslint-plugin-react-hooks/tsconfig.json',
  },
```

poteto reviewed (approved):
Thanks!
